server/src/main/java/org/opensearch/index/SegmentReplicationPressureService.java

```diff
@@ -154,4 +181,70 @@ public void setMaxAllowedStaleReplicas(double maxAllowedStaleReplicas) {
     public void setMaxReplicationTime(TimeValue maxReplicationTime) {
         this.maxReplicationTime = maxReplicationTime;
     }
+
+    @Override
+    public void close() throws IOException {
+        failStaleReplicaTask.close();
+    }
+
+    // Background Task to fail replica shards if they are too far behind primary shard.
+    final class AsyncFailStaleReplicaTask extends AbstractAsyncTask {
+
+        AsyncFailStaleReplicaTask(TimeValue interval) {
+            super(logger, threadPool, interval, true);
+            rescheduleIfNecessary();
+        }
+
+        @Override
+        protected boolean mustReschedule() {
+            return true;
+        }
+
+        @Override
+        protected void runInternal() {
+            if (isSegmentReplicationBackpressureEnabled) {
+                final SegmentReplicationStats stats = tracker.getStats();
+                for (Map.Entry<ShardId, SegmentReplicationPerGroupStats> entry : stats.getShardStats().entrySet()) {
+                    final Set<SegmentReplicationShardStats> staleReplicas = getStaleReplicas(entry.getValue().getReplicaStats());
+                    final ShardId shardId = entry.getKey();
+                    final IndexService indexService = indicesService.indexService(shardId.getIndex());
+                    final IndexShard primaryShard = indexService.getShard(shardId.getId());
+                    for (SegmentReplicationShardStats staleReplica : staleReplicas) {
+                        if (staleReplica.getCurrentReplicationTimeMillis() > 2 * maxReplicationTime.millis()) {
```

Comment:
What would the default value be for max replication time? Is there any documentation we're providing on how to set this value when people enable backpressure?

Reply:
the default value for [max replication time](https://github.com/opensearch-project/OpenSearch/blob/7b708c45558e9557d7306833e013eb8e6d00c5ec/server/src/main/java/org/opensearch/index/SegmentReplicationPressureService.java#L58) is 5min. For documentation, we need to open an issue with documentation repo documenting all changes/usages of segrep backpressure. I will open that issue after this PR gets merged in.